flowmachine/flowmachine/features/subscriber/active_subscribers.py

```diff
@@ -0,0 +1,170 @@
+        The number of major periods a subscriber must appear active in to appear in the output
+        of the query
+    period_unit: {'days','hours','minutes'} default 'hours'
+        The unit of time to of minor_period_length
+    subscriber_identifier : {'msisdn', 'imei'}, default 'msisdn'
+        Either msisdn, or imei, the column that identifies the subscriber.
+    subscriber_subset : str, list, flowmachine.core.Query, flowmachine.core.Table, default None
+        If provided, string or list of string which are msisdn or imeis to limit
+        results to; or, a query or table which has a column with a name matching
+        subscriber_identifier (typically, msisdn), to limit results to.
+    hours: tuple default None
+        A range of hours to restrict contributions to minor_period_counts to.
+
+    Notes
+    -----
+    * if minor_period_length is equal to or less than the length of hours, there will be a set
+    of major_periods that are guaranteed to be empty (as they fall outside the range set by hours).
+
+    Examples
+    --------
+    Returns subscribers who were active for at least one hour a day for at least three days
+    between 2016-01-01 and 2016-01-04
+
+    >>>    active_subscribers = ActiveSubscribers(
+                start_date=date(year=2016, month=1, day=1),
+                minor_period_length=1,
+                minor_periods_per_major_period=24,
+                total_major_periods=4,
+                minor_period_threshold=1,
+                major_period_threshold=3,
+                tables=["events.calls"],
+            )
+
+    Returns subscribers that were active in at least two ten minute intervals within half an hour,
+    at least three times across the two hours between 20:00:00 and 22:00:00 on 2016-01-01
+
+    >>>     active_subscribers = ActiveSubscribers(
+                start_date=datetime(year=2016, month=1, day=1, hour=20),
+                minor_period_length=10,
+                minor_periods_per_major_period=3,
+                total_major_periods=4,
+                minor_period_threshold=2,
+                major_period_threshold=3,
+                period_unit="minutes",
+                tables=["events.calls"],
+            )
+
+
+    """
+
+    period_to_rrule_mapping = {
+        "days": rr.DAILY,
+        "hours": rr.HOURLY,
+        "minutes": rr.MINUTELY,
+    }
+
+    def __init__(
+        self,
+        start_date: Union[date, datetime, str],
+        minor_period_length: int,
+        minor_periods_per_major_period: int,
+        total_major_periods: int,
+        minor_period_threshold: int,
+        major_period_threshold: int,
+        period_unit: Literal["days", "hours", "minutes"] = "hours",
+        subscriber_identifier: Optional[str] = "msisdn",
+        tables: Optional[List[str]] = None,
+        subscriber_subset: Optional[SubscriberSubsetType] = None,
+        hours: Optional[Tuple[int, int]] = None,
+    ):
+        self.start_date = start_date
```

Comment:
Given that `ExposedDatetimeMixin` is used here (so this query does have an `end_date` attribute), we could set the end date attribute to the value `start_date + (period_unit * minor_period_length * minor_periods_per_major_period * total_major_periods)`.